packages/casting/src/follower-cosmjs.js

```diff
@@ -112,290 +138,376 @@ export const makeCosmjsFollower = (
-      );
-      assert(
-        dataPrefixBytes.every((v, i) => v === result[i]),
-        X`${result} doesn't start with data prefix ${dataPrefixBytes}`,
-      );
-      return result.slice(dataPrefixBytes.length);
-    };
+  /**
+   * @param {number} blockHeight
+   */
+  const tryGetDataAtHeight = blockHeight => {
+    const getProvenData = () => getProvenDataAtHeight(blockHeight);
+    const getUnprovenData = () => getUnprovenDataAtHeight(blockHeight);
+    return queryVerifier(getProvenData, crash, getUnprovenData);
+  };
 
-  const getProvenValueAtHeight = makeQuerier('queryVerified');
-  const getUnprovenValueAtHeight = makeQuerier('queryUnverified');
+  /**
+   * @param {number} blockHeight
+   */
+  const getDataAtHeight = async blockHeight => {
+    for (let attempt = 0; ; attempt += 1) {
+      try {
+        // AWAIT
+        return await tryGetDataAtHeight(blockHeight);
+      } catch (error) {
+        // We expect occasionally to see an error here if the chain has not
+        // reached the requested blockHeight.
+        await E(leader).retry(where, error, attempt);
+        continue;
+      }
+    }
+  };
 
   /**
-   * @param {IterationObserver<FollowerElement<T>>} updater
-   * @param {AsyncIterable<FollowerElement<T>>} iterable
+   * @param {number} blockHeight
+   * @param {Uint8Array} data
    */
-  const getIterable = async (updater, iterable) => {
-    let finished = false;
+  const streamCellForData = (blockHeight, data) => {
+    const cell = decode(data);
+
+    if (
+      cell &&
+      Array.isArray(cell.values) &&
+      typeof cell.blockHeight === 'string'
+    ) {
+      blockHeight = Number(cell.blockHeight);
+      if (!Number.isNaN(blockHeight)) {
+        return {
+          blockHeight: Number(cell.blockHeight),
+          values: cell.values,
+        };
+      }
+    }
 
-    const fail = err => {
-      finished = true;
-      updater.fail(err);
-      return false;
+    // Coerce legacy value cells to stream cells at their given height.
+    // Since this is either the first iteration or the data varies bytewise
+    // from the data on the previous block, we can assume the blockHeight is
+    // the current block.
+    return {
+      blockHeight,
+      values: [cell],
     };
+  };
 
-    const crash = err => {
-      fail(err);
-      if (crasher) {
-        E(crasher)
-          .crash(`PROOF VERIFICATION FAILURE; crashing follower`, err)
-          .catch(e => assert(false, X`crashing follower failed: ${e}`));
-      } else {
-        console.error(`PROOF VERIFICATION FAILURE; crashing follower`, err);
+  /**
+   * @param {number} blockHeight
+   */
+  const getStreamCellAtHeight = async blockHeight => {
+    for (;;) {
+      // AWAIT
+      const data = await getDataAtHeight(blockHeight);
+      if (data.length !== 0) {
+        const streamCell = streamCellForData(blockHeight, data);
+        return { data, streamCell };
       }
-    };
+      // AWAIT
+      await E(leader).jitter(where);
```

Comment:
This retry loop will never exit. I need to account for the case of an empty data cell differently here, and recover in the caller. We previously allowed JSON.parse to gracelessly throw if we subscribed before the first publish and I think we can do better by waiting for a block to produce some data.